ops/model.py

```diff
@@ -1326,6 +1327,145 @@ def list_files(self, path: str, *, pattern: str = None,
         """
         return self._pebble.list_files(path, pattern=pattern, itself=itself)
 
+    def push_path(self,
+                  source_path: typing.Union[typing.Union[str,
+                                                         Path],
+                                            typing.List[typing.Union[str,
+                                                                     Path]]],
+                  dest_dir: typing.Union[str,
+                                         Path]):
+        """Push all the things.
+
+        TODO: write this docstring
+        """
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        def local_list(source_path: str):
+            paths = [source_path]
+            if os.path.isdir(source_path):
+                paths = os.listdir(source_path)
+            files = []
+            for name in paths:
+                fpath = os.path.join(source_path, name)
+                info = os.lstat(fpath)
+                if stat.S_ISDIR(info.st_mode):
+                    ftype = pebble.FileType.DIRECTORY
+                elif stat.S_ISREG(info.st_mode):
+                    ftype = pebble.FileType.FILE
+                elif stat.S_ISLINK(info.st_mode):
+                    ftype = pebble.FileType.SYMLINK
+                else:
+                    continue
```

Comment:
This feels like we should either complete it (`socket`, `named-pipe`, `device`, `unknown`) or log a warning about that particular file (I doubt if we want to push any of these anyway) and move on.

I don't see a negative assertion in the tests, but I would guess that bad things happen if we try to push `/dev/kmem` or something. 

Considering the usage below (`pebble.FileInfo(..., ftype, stat.S_IMODE(info.st_mode))`) is disconnected from the logic here, it may merit slapping these into a dict, `NamedTuple`, or some other 1:1 mapping for clarity.

Reply:
I think I'm going to avoid the dict seems a bit overkill for this.  Good call on logging the skipped files - that's added.  And I'm still working on more tests and robust error handling - hence the "draft" status :P